esp-hal/src/dma/mod.rs

```diff
@@ -397,6 +399,24 @@ macro_rules! dma_circular_descriptors {
     };
 }
 
+/// Allocates a DMA buffer with a specific size, aligned to 4 bytes
+#[doc(hidden)]
+#[macro_export]
+macro_rules! declare_aligned_dma_buffer {
+    ($name:ident, $size:expr) => {
+        static mut $name: [u32; ($size + 3) / 4] = [0; ($size + 3) / 4];
```

Comment:
```suggestion
        static mut $name: [u32; $size.div_ceil(4)] = [0; $size.div_ceil(4)];
```

I think you can also do something like this to get alignment with less `unsafe`.
```Rust
#[repr(align(4))]
struct Data([u8; $size.div_ceil(4)]);
static mut $name: Data = ....;
name.0
```

Reply:
"Can't call `div_ceil` on an ambiguous numeric type" or something similar

Reply:
Also would complain about redefining Data

Reply:
> Also would complain about redefining Data

True, might be worth making a type for it in the hal. And users could use it in other scenarios, like with `alloc`.

Bah, the numeric type thing is annoying.

Reply:
I could just cast to usize but it'd just end up being more code than manual rounding up.

Reply:
what about [aligned](https://crates.io/crates/aligned)?

Reply:
I'll invoke my oneliner rule against aligned. We don't use that anywhere else except in examples, so I'll not introduce a new dependency to rework a oneliner.

Reply:
On second thought, I think it might be better to not "lie" to the user about the size here and assert (at compile time?) that `$size % 4 == 0` (only on the PDMA chips if possible), rather than discretely alter the size for the user.
Better for a user's program to not compile than to subtly/silently send/receive more bytes than intended.
Having said that, changing the size of the returned buffer (or asserting) is a breaking change.

Reply:
Technically, we shouldn't end up sending/receiving more data, all we are doing here is making sure nothing gets placed into the 2 bytes that may be written out of bounds. What the user and the HAL ends up seeing is the same sized buffer they asked for.

I agree that this is a terrible hack to solve the problem, I just don't know any better solutions. Fun fact: the DMA works as expected for 1 mod 4 buffers, which makes me think this might be a hardware issue. But I admit I stopped reading the TRM after only 2 hours or so, hunting for a clue, so I may have missed the "auto-burst mode if you transfer more than 1 byte" bit in one of the registers.